google/auth/credentials.py

```diff
@@ -236,6 +223,40 @@ def with_non_blocking_refresh(self):
         self._use_non_blocking_refresh = True
 
 
+class CredentialsWithTrustBoundary(Credentials):
+    """Abstract base for credentials supporting trust boundary factory"""
+
+    def apply(self, headers, token=None):
+        super().apply(headers, token)
+        """Trust boundary value will be a cached value from global lookup.
+
+        The response of trust boundary will be a list of regions and a hex
+        encoded representation.
+
+        An example of global lookup response:
+        {
+            "locations": [
+                "us-central1", "us-east1", "europe-west1", "asia-east1"
+            ]
+            "encoded_locations": "0xA30"
+        }
+        """
+        if self._trust_boundary is not None:
+            headers["x-allowed-locations"] = self._trust_boundary["encoded_locations"]
+
+    def lookup_trust_boundary(self, request):
+        """Lookup trust boundary shall be implemented by subclasses"""
+        raise NotImplementedError("Missing definition of trust boundary lookup")
+
+    def set_trust_boundary(self, trust_boundary):
+        """Sets the trust boundary value to the credential"""
+        self._trust_boundary = trust_boundary
```

Comment:
Use a python property instead: https://stackoverflow.com/questions/6618002/using-property-versus-getters-and-setters